src/data/bucket/line_bucket.js

```diff
@@ -319,16 +319,24 @@ class LineBucket implements Bucket {
         const cap = layout.get('line-cap').evaluate(feature, {});
         const miterLimit = layout.get('line-miter-limit');
         const roundLimit = layout.get('line-round-limit');
+        const dashArray = this.layers[0].paint.get('line-dasharray').value;
+        const pattern = this.layers[0].paint.get('line-pattern').value;
+
+        const hasDashArray = dashArray.kind !== 'constant' || !!dashArray.value;
+        const hasGradient = !!this.layers[0].paint.get('line-gradient');
+        const hasPattern = pattern.kind !== 'constant' || !!pattern.value;
+        const adjustSharpCorners = hasDashArray || hasGradient || hasPattern;
+
         this.lineClips = this.lineFeatureClips(feature);
 
         for (const line of geometry) {
-            this.addLine(line, feature, join, cap, miterLimit, roundLimit);
+            this.addLine(line, feature, join, cap, miterLimit, roundLimit, adjustSharpCorners);
```

Comment:
🤔 I think there is an architectural problem with the approach. In most of the cases, bucket's data that is created at layout phase, can't depend on paint properties. Can be tested by creating line layer and then setting dash array or gradient paint property using runtime apis.

Reply:
Thanks for pointing this out, @alexshalamov! That makes perfect sense, though I do observe that this layout code is re-executed when I change the `line-dasharray`, e.g.:

```
map.setPaintProperty('path', 'line-dasharray', [2, 1])
```

Reply:
Ah, now I see. I think my success was dumb luck, as I only tested this situation with `line-dasharray`. My current hypothesis is that the data-driven nature of `line-dasharray` causes re-execution of the line bucket layout code, which was sufficient to trigger this to update. As `line-gradient` is *not* data-driven, the gradient paint property does not trigger the update and the state is not correctly updated.

In summary, I think you're right that this is a fundamental error in my approach. I think it's necessary to simply close this PR since it's just too ugly of a hack which, more importantly, doesn't cover all cases.